### PR TITLE
fix(cosh-ng): [shell] fd dup redirect is not write

### LINKED ISSUE #2054: [TestAgent][cosh-ng] bug: 2>&1 被 fail-closed 分类为 write redirection,纯只读命令被误报为高危
### Component

cosh-ng

### Bug Description

命令风险分类器对 `2>&1`(stderr → stdout fd duplication,纯只读)走 fail-closed 路径,统一标为 `RedirectionWrite` → high risk / write redirection。导致所有带 `2>&1` 合并错误的只读命令(如 `ls`、`cat`、`find`、`command --version`)都被误报为高危,需要用户手动 Allow 才能执行。

### Steps to Reproduce

```shell
ssh root@<ECS-IP>   # 授权测试机,读者自行替换
cosh
# 在交互界面内问 agent 一个会合并 stderr 的只读请求,例如:
#   帮我看一下 /usr/share/anolisa/runtime/skills/ws-ckpt/ 这个 skill 装了没?
# agent 派发的 bash tool_use 形如:
#   rtk ls -la /usr/share/anolisa/runtime/skills/ws-ckpt/ 2>&1 && echo "---" && rtk ls -la ~/.cosh/skills/ws-ckpt/ 2>&1
```

实测出现的 approval 面板:

```
╭ Approval req-7 ─────────────────────────────────────────────────────────────────────────
│ Bash · high risk
│ └ Risk: write redirection
│ $ rtk ls -la /usr/share/anolisa/runtime/skills/ws-ckpt/ 2>&1 && echo "---" && rtk ls -la ~/.cosh/skills/ws-ckpt/ 2>&1
│ > [ Allow once ]     [ Always trust ]     [ Deny ]     [ Details ]
╰────────────────────────────────────────────────────────────────────────────────────────
```

### Actual Behavior

- `rtk ls -la <path> 2>&1` 被分类为 `RiskImpact::High`,reason 含 `redirection-write`
- approval 面板显示 `Bash · high risk` + `Risk: write redirection`
- 用户必须手动 Allow 一次或 Always trust 才能执行
- 命令本身是纯只读(`ls -la` + `2>&1` 合并错误流),没有任何文件写操作

### Expected Behavior

- `2>&1` 是 fd duplication,不是 filesystem write,不应被标为 `RedirectionWrite` shape
- 纯只读命令 + `2>&1` 应保持原命令本身的风险级别(`ls` / `cat` / `find` 等只读命令应为 medium 或 low)
- 只有真正写文件的 `>` / `>>` 形式(`echo > file`、`cat foo 2> bar`)才应标为 write redirection

#### 1. `>` 分支的 lookahead 在遇到 `&` 时 break,target 解析为空

`src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs:110-175` 处理 `>` 的 lookahead 时,在 target 字符收集循环里把 `&` 列为 break 字符:

```rust
while let Some(&next) = lookahead.peek() {
    if matches!(
        next,
        ' ' | '\t' | '\n' | ';' | '|' | '&' | '<' | '>' | '(' | ')' | '{' | '}'
    ) {
        break;
    }
    // ...
}
```

`command_risk_parser.rs:143-147` 的 break set 包含 `'&'`,所以 `2>&1` 的 target 解析到 `&` 就 break,target = `""`(空字符串)。

#### 2. 空字符串不在白名单 `SAFE_OUTPUT_SINKS`

`src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs:11`:

```rust
const SAFE_OUTPUT_SINKS: &[&str] = &["/dev/null"];
```

`SAFE_OUTPUT_SINKS.contains("")` → false,所以走 else 分支:

```rust
if !guarded && literal && SAFE_OUTPUT_SINKS.contains(&target.as_str()) {
    // ... null_redirections += 1 (降级路径)
} else {
    if fd_candidate {
        push_token(&mut tokens, &mut token, &mut token_quoted);
    }
    if chars.peek().is_some_and(|next| *next == '>') {
        chars.next();
    }
    shape = max_shape(shape, CommandShape::RedirectionWrite);  // ← bug 在这
}
```

`command_risk_parser.rs:173` 把 shape 升到 `RedirectionWrite`。

#### 3. `RedirectionWrite` shape 在 `command_risk.rs` 直接走 high risk 分支

`src/cosh-ng/crates/cosh-shell/src/tools/command_risk.rs:75-79`:

```rust
if parsed.shape == CommandShape::RedirectionWrite {
    // ... return high risk assessment with reason "redirection-write"
}
```

#### 4. 设计取舍来自 issue #1667,只把 `/dev/null` 加白名单

`src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs:3-10` 注释明确写了设计边界:

```rust
/// Non-persistent output-suppression sink allowlist (issue #1667
/// implementation boundaries): a `[N]>` / `[N]>>` redirection is treated
/// as output suppression instead of a filesystem write only when the
/// target is an unquoted, non-expanded literal from this table. Every
/// other form (regular files, quoted or expanded targets, `2>&1`, `&>`)
/// keeps the fail-closed RedirectionWrite high-risk path.
```

issue #1667 的边界只允许 `/dev/null` 作为 sink,其他形式(包括 `2>&1` fd duplication)都 fail-closed。理由是"parser 无法可靠区分 `2>&1` 和 `2>file`",但实际上这俩语法有明显区别:`2>&1` 的 `&` 表示 fd 别名,`2>file` 没有 `&`,target 是文件路径。

#### 5. 测试 V-M9 锁死了这个 by-design 行为

cosh-shell 风险分类测试集里的 V-M9 用例显式断言 `ls 2>&1` 必须是 `RiskImpact::High` + reason `redirection-write`,且不含 `output-suppressed`。修复本 issue 需要同步更新 V-M9 用例。

#### A. 在 `>` 分支的 lookahead 里识别 `>&N` fd duplication(推荐,最直接)

修改 `src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs:110-175` 的 `>` 处理分支,在 target 解析前先 check `>&N` 形式:

```rust
'>' => {
    let guarded = amp_redirect_guard;
    amp_redirect_guard = false;
    let fd_candidate = !token.is_empty()
        && !token_quoted
        && token.bytes().all(|byte| byte.is_ascii_digit());
    if !fd_candidate {
        push_token(&mut tokens, &mut token, &mut token_quoted);
    }

    // fd duplication check: '>&' followed by digit or '-' (e.g., 2>&1, 2>&2, 2>&-)
    // This is NOT a filesystem write — it aliases one fd to another.
    let mut fd_dup_lookahead = chars.clone();
    let mut fd_dup_consumed = 0usize;
    let mut is_fd_dup = false;
    if fd_dup_lookahead.peek().is_some_and(|next| *next == '&') {
        fd_dup_lookahead.next();
        fd_dup_consumed += 1;
        if fd_dup_lookahead
            .peek()
            .is_some_and(|next| next.is_ascii_digit() || *next == '-')
        {
            fd_dup_lookahead.next();
            fd_dup_consumed += 1;
            is_fd_dup = true;
        }
    }

    if is_fd_dup {
        // Consume the &N (or &-) and do NOT set RedirectionWrite shape.
        for _ in 0..fd_dup_consumed {
            chars.next();
        }
        // The fd prefix (e.g., "2") is consumed as part of the redirection
        // syntax, not pushed as an ordinary argument.
        token.clear();
        token_quoted = false;
        continue;
    }

    // ... existing lookahead logic for file targets (unchanged)
}
```

#### B. 调整 V-M9 测试用例

修改 cosh-shell 风险分类测试集里的 V-M9,把 `ls 2>&1` 从 "High + redirection-write" 改成 "Medium(或 Low)+ 不含 redirection-write"。`ls &>/dev/null` 保留原行为(`&>` 是合并 stdout+stderr 写文件,仍应 high)。

#### C. 同时识别 `>&` 不跟数字的形式(可选,边界保护)

`>&` 不跟数字(`>&file`)在某些 shell 等价于 `>file 2>&1`,这是真写文件,应保持 high。修复 A 不影响这条路径,因为 A 的判断要求 `&` 后必须跟数字或 `-`;`>&file` 形式会落入原有逻辑,target 解析为 `file`,不在白名单,继续走 RedirectionWrite 分支。

### Environment

- OS: Anolis OS 8.x
- Component: cosh-ng
- 测试时间: 2026-07-31

### Additional Context

- 与 #2045(`/hooks --help` 不列子命令)、#2048(抢答 stale data)、#2052(recovery 频繁触发导致 cache 失效)均无关。
- issue #1667 是 `/dev/null` 白名单的来源,本 issue 是在 #1667 边界外扩展识别 `>&N` fd duplication。
- 稳定性:行为已被 V-M9 测试锁死,非 flaky;修复需同步更新测试。
- 影响面:任何带 `2>&1` 合并错误流的只读命令都会被误报。常见场景包括 `ls -la <path> 2>&1`、`cat <file> 2>&1`、`find . -name ... 2>&1`、`command --version 2>&1` 等纯读操作。

### PR DESCRIPTION
**Pull request by SunnyQjm**
# Summary

`2>&1`-style fd duplication is parsed with an empty redirection target, misses the `/dev/null` allowlist, and fails closed to the `RedirectionWrite` shape, so every read-only command carrying `2>&1` is reported as **high risk / write redirection** on the approval card and requires a manual Allow (issue #2054; boundary inherited from issue #1667, whose spec explicitly deferred fd duplication to a follow-up).

This PR exempts by policy the **conventional output targets** — `[N]>&1` / `[N]>&2` (duplication) and `[N]>&-` (close) — from the write-redirection escalation. The command then keeps the risk of its real remaining shape (e.g. `ls 2>&1` → medium AskUser instead of high). The policy rationale (review rounds 6–8): rebinding fd 1/2 (`exec 2>out`) is possible in the persistent foreground shell, but that rebinding command is itself fail-closed High and user-approved first, and once rebound the file receives output from commands with no redirection syntax at all — a lexical classifier cannot defend that state either way. Arbitrary numeric targets (`>&3`, `2>&10`) stay fail-closed: their bindings are only reachable through explicit fd syntax, where lexical rejection is effective. Bash's move form `[N]>&M-` is not elided either: zsh has no move syntax and really writes a file named `M-` (review round 2).

# Changes

- `crates/cosh-shell/src/tools/command_risk_parser.rs`: fd word probe at the head of the `'>'` arm.
  - Recognition requires **all** of: not `&>`-led (`amp_redirect_guard`), not `>>`, a single-digit-or-absent source prefix (zsh treats only a lone digit as an fd — `tee 10>&1` passes `10` as an argument and creates a file named `10`, review round 5), `&` followed by exactly `1`, `2`, or `-` (conventional output targets only, review round 6), and a word boundary right after. The boundary set is whitespace plus POSIX operator characters (`;|&<>()`) only — `{`/`}` are not operators in the redirection-word position (both bash and zsh write a file named `1{` for `>&1{`, review round 3) and are not boundaries. Anything else falls through **byte-for-byte** to the pre-existing fail-closed path, so every blind spot stays conservative (mis-conservative, never mis-permissive).
  - On recognition the IO_NUMBER prefix (`2` in `2>&1`) and the fd word are consumed as redirection syntax (no argv leakage) and the shape is not raised. Duplications add no reason; closes of an **output stream** (bare default, fd 1, fd 2) hide user-visible output and join the issue #1667 null-sink channel (`output-suppressed`, never auto-allowed; review round 1), while `0>&-`/`3>&-` keep visible output untouched and elide without annotation (review round 4).
- `crates/cosh-shell/src/tools/command_risk_tests.rs`: decision-matrix tests (elision positives incl. `>&2`, trailing-space `ls 2>&1 `, `0>&-`/`3>&-`; suppression-annotated closes `2>&-`/`1>&-`/`>&-`; fail-closed lookalikes incl. `2>&1x`, `2>&$FD`, `2>&1-`, `>&1{`, `>&1}`, `>&-{`, `tee 10>&1`, `10>&-`, `printf payload >&3`, `2>&10`, `2>&3`, `>&file`, `&>&1`, `2>& 1`, `2>>&1`), invariant tests (argv preservation incl. zsh-semantics `10`, `&&` boundary, high-risk tails never masked), and execution-boundary anchors (three auto-allow hard gates keep rejecting the raw `>`; `ls 2>&1` is never AutoAllow).
  - Re-anchored: V-M9 `ls 2>&1` moved to the elision cases (`ls &>/dev/null` stays High); design v3 §1 `echo hi >&2` moved likewise (`ls >| out.txt`, `ls > /dev/nul*` unchanged).
- `scripts/check-test-inventory.sh`: cosh-shell exact lib/bin overlap ceiling 646 → 651 (five new tests in the shared tools module; measured via `cargo test -- --list` on both targets).

# Tests

- New: `fd_duplication_is_not_redirection_write`, `fd_duplication_compound_and_pipeline_keep_segment_assessment`, `fd_duplication_lookalikes_fail_closed_to_high`, `fd_duplication_execution_boundary_stays_closed`, `fd_duplication_and_null_suppression_compose_independently`.
- Review-round regressions: `2>&1-` (zsh writes `1-`), `>&1{`/`>&1}`/`>&-{` (`{}` not a boundary; both shells write `1{`), close forms annotated `output-suppressed`, trailing-space boundary case.
- FAIL→PASS: pre-fix probe on base `e8a9b8ff` asserts `ask("ls 2>&1")` is not High → FAILED; same assertions pass post-fix (now the permanent regression tests above).

# Verification

- Focused (macOS host, worktree): `cargo test -p cosh-shell --lib` 1179 passed / 0 failed; `--bin cosh-shell` 2205 passed / 0 failed / 1 ignored (pre-existing); `cargo clippy -p cosh-shell --all-targets` clean; `cargo fmt --check` clean.
- CI gates run locally: `check-layout.sh` pass, `check-test-inventory.sh` pass (ceiling updated to measured 651).
- Real-machine acceptance (alinux3 arm64 container, real `cosh-shell raw` + real `cosh-core` headless + real provider): agent-dispatched `ls -la /tmp 2>&1` now surfaces an approval card reading **`Bash · 中风险`** with the command verbatim and no write-redirection reason; after Allow the command executes (exit 0) and the turn completes. Scripted-PTY case green, evidence verify passed.
- Excluded scope (not run): full workspace tests on linux; the raw_cli signal-semantics suite fails identically on base and on this branch in the local container backend (no init/zombie reaping; A/B verified) and defers to GitHub CI as the authority.

# Evidence

Real-machine acceptance (alinux3 arm64 container, real cosh-core adapter over a real PTY, scripted case FDDUP-APPR-001; screenshots derived from the sanitized cast of this branch's binaries):

**Approval card after the fix — `Bash · 中风险` (medium), command verbatim, no write-redirection reason** (before the fix this exact shape rendered `Bash · high risk` + `Risk: write redirection`, see issue #2054):

approval card medium risk

**Final state — command approved, executed (exit 0), agent summarized, session exited cleanly:**

final state

Closes #2054

### CODE REVIEW

File: src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs [hunk 1]
```diff
@@ -122,6 +125,61 @@ pub(super) fn parse_command(command: &str) -> ParsedCommand {
...
+                    while dup_lookahead
+                        .peek()
+                        .is_some_and(|next| next.is_ascii_digit())
+                    {
+                        dup_lookahead.next();
+                        dup_consumed += 1;
+                        has_digit = true;
+                    }
+                    let mut has_dash = false;
+                    if dup_lookahead.peek().is_some_and(|next| *next == '-') {
```
Comment (kongche-jbw) [inline src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs]: [P1] zsh 下该分支会漏报真实文件写入

这里接受数字后可选 `-`，把 `[N]>&M-` 当作 fd move 并跳过 `RedirectionWrite`。但 cosh-shell 会把批准命令交给当前前台 shell，且支持 zsh；zsh 5.9 没有 Bash 的 move 语法，`zsh -f -c '{ print -u2 ERR; } 2>&1-'` 会创建或截断文件 `1-`。因此 zsh 用户会看到非高危审批，实际却发生文件写入。建议本 PR 只豁免跨 shell 一致的 `[N]>&M` 和 `[N]>&-`，让 `[N]>&M-` 继续 fail-closed；若后续支持 move，需要带 shell 类型解析。

File: src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs [hunk 2]
```diff
@@ -122,6 +125,61 @@ pub(super) fn parse_command(command: &str) -> ParsedCommand {
...
+                        dup_lookahead.next();
+                        dup_consumed += 1;
+                        has_dash = true;
+                    }
+                    let boundary_ok = dup_lookahead.peek().is_none_or(|next| {
```
Comment (kongche-jbw) [inline src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs]: [P1] 保持 Rust 1.74 MSRV

`Option::is_none_or` 直到 Rust 1.82 才稳定，而 workspace、README 和 CONTRIBUTING 均声明最低 Rust 版本为 1.74；使用 1.74 构建会在此处失败。当前 stable CI 无法发现该回归。请改为 `map_or(true, ...)`，必要时沿用仓库现有的窄范围 Clippy allow。

**Review by kongche-jbw — CHANGES_REQUESTED** [review]
请求修改：存在两个正确性问题：项目支持的 zsh 会将 [N]>&M- 作为真实文件重定向，但当前实现将其分类为非写入；同时 Option::is_none_or 破坏了文档声明的 Rust 1.74 最低版本。除此之外，基于 stable 工具链的范围内测试、Clippy、格式检查、布局审计和测试清单均通过。

File: src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs [hunk 3]
```diff
@@ -122,6 +125,68 @@ pub(super) fn parse_command(command: &str) -> ParsedCommand {
...
+                        dup_consumed += 1;
+                        has_dash = true;
+                    }
+                    // Keep the workspace Rust 1.74 MSRV; `Option::is_none_or`
+                    // is newer.
+                    #[allow(clippy::unnecessary_map_or)]
+                    let boundary_ok = dup_lookahead.peek().map_or(true, |next| {
+                        matches!(
+                            next,
+                            ' ' | '\t' | '\n' | ';' | '|' | '&' | '<' | '>' | '(' | ')' | '{' | '}'
```
Comment (kongche-jbw) [inline src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs]: [P1] 不要把 `{` / `}` 当作 fd duplication 词边界

这里将 `{` 和 `}` 列为 boundary，会把 `>&1{` 拆成 fd duplication 加 `Complex`，从而跳过 `RedirectionWrite`。但 `{`/`}` 不是这种位置上的 shell 元字符边界：在隔离目录中，`bash -c ': >&1{'` 与 `zsh -f -c ': >&1{'` 都以 0 退出并创建文件 `1{`。因此当前实现仍会把真实文件写入显示为非高危。请从 fd duplication 的边界集合中移除 `{`/`}`，并增加 `>&1{`、`>&1}`、`>&-{` 等 fail-closed 回归用例。

File: src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs [hunk 4]
```diff
@@ -3,9 +3,12 @@ use super::command_risk::CommandShape;
 /// Non-persistent output-suppression sink allowlist (issue #1667
 /// implementation boundaries): a `[N]>` / `[N]>>` redirection is treated
 /// as output suppression instead of a filesystem write only when the
-/// target is an unquoted, non-expanded literal from this table. Every
-/// other form (regular files, quoted or expanded targets, `2>&1`, `&>`)
-/// keeps the fail-closed RedirectionWrite high-risk path. Extending this
+/// target is an unquoted, non-expanded literal from this table. Fd
+/// duplication forms (`[N]>&M`, `[N]>&M-`, `[N]>&-`) are recognized
```
Comment (kongche-jbw) [inline src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs]: [P2] 同步清理已排除 fd move 的旧说明

最新实现和测试已经明确让 `[N]>&M-` fail-closed，但这里仍宣称该形式会被识别为非文件写入；PR body 与提交主体也仍把 `2>&1-` 列为 elision 正例。这与下方实现注释及回归测试相矛盾，容易让后续维护者重新引入刚修复的跨 shell 漏洞。请统一改为仅识别 `[N]>&M` 和 `[N]>&-`，并把 `[N]>&M-` 明确记录为非目标/fail-closed 形式。

**Review by kongche-jbw — CHANGES_REQUESTED** [review]
继续复核发现一个新的阻塞性误判和一处文档漂移：`{`/`}` 被错误地当作 fd duplication 词边界，仍可把真实文件写入降级为非写入；同时 fd move 已被排除，但顶部注释、PR 描述和提交主体尚未同步。前一轮两条 P1 已验证修复，范围内测试、Clippy、格式检查、布局审计、测试清单及当前 cosh-ng CI 均通过。

File: src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs [hunk 5]
```diff
@@ -122,6 +127,74 @@ pub(super) fn parse_command(command: &str) -> ParsedCommand {
...
+                            token_quoted = false;
+                        }
+                        // Pure close forms (`[N]>&-`) suppress the stream
+                        // entirely from the user's point of view, so they
+                        // join the issue #1667 null-sink channel
+                        // (`output-suppressed` reason + auto-allow
+                        // fallback). Duplications (`2>&1`) keep the stream
+                        // visible and stay annotation-free (PR #2069
+                        // review P1).
+                        if has_dash {
```
Comment (kongche-jbw) [inline src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs]: [P2] 想请确认一下，这里是否需要区分被关闭的 fd。当前任意 `[N]>&-` 都会进入 `null_redirections`，因此像 `0>&-`（关闭 stdin）和 `3>&-`（关闭辅助 fd）也会得到 `output-suppressed`；但在 bash、zsh 中，`printf visible 0>&-` 和 `printf visible 3>&-` 都仍会正常输出 `visible`。对比基线后确认，这个具体的理由标记是本 PR 新增的行为（基线会保守落入 `RedirectionWrite`），不过它不涉及新的自动执行放宽。是否可以考虑只让默认 fd、1、2 进入现有输出抑制通道，或为其他 fd 使用更中性的 `descriptor-closed` 理由？

**Review by kongche-jbw — COMMENTED** [review]
继续审查补充一个非阻断的理由语义问题，想请作者确认一下是否需要区分被关闭的 fd。

File: src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs [hunk 6]
```diff
@@ -122,6 +127,80 @@ pub(super) fn parse_command(command: &str) -> ParsedCommand {
...
+                        }
+                        // Close forms (`[N]>&-`) suppress the stream from
+                        // the user's point of view only when they close an
+                        // output stream: the bare default (stdout), fd 1,
+                        // or fd 2. Closing stdin or an auxiliary fd
+                        // (`0>&-`, `3>&-`) leaves the visible output
+                        // untouched (verified in bash and zsh), so those
+                        // stay annotation-free like duplications
+                        // (PR #2069 review rounds 1 and 4).
+                        let closes_output_stream =
```
Comment (kongche-jbw) [inline src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs:207]: [P1] 想请再确认一下多位“源 fd 前缀”的跨 shell 语义。zsh 5.9 的重定向规则只把操作符前的单个数字视为 fd；实测 `tee 10>&1` 在 bash 中把 `10` 当作 fd 10、不会创建文件，但在 zsh 中会把 `10` 保留为 `tee` 的普通参数并创建文件 `10`。当前 `fd_candidate` 接受任意长度的纯数字，随后这里会按 fd 语法清除该 token，因此 zsh 实际使用的文件参数会从风险评估 argv 中消失，并把基线的 High 降为非 High。`printf visible 10>&-` 也体现同一差异：bash 关闭 fd 10、输出仍可见，zsh 则保留参数 `10` 并关闭默认 stdout。考虑到 cosh-shell 会执行当前前台 zsh，是否可以让多位源前缀继续 fail-closed，或按 shell 类型解析？现有 `2>&10` 覆盖的是多位目标 fd，与这个源前缀场景不同；建议补充 `tee 10>&1` / `10>&-` 的跨 shell 回归。

**Review by kongche-jbw — CHANGES_REQUESTED** [review]
感谢持续修订，上一轮关于 `0>&-` / `3>&-` 的问题已经正确处理。继续做 bash/zsh 交叉验证时，又发现一个多位“源 fd 前缀”的差异，可能还需要再收紧一下；具体见行内 P1。

另外有一处非阻断的 [P2] 文档同步项：PR 描述的 Changes 仍写成所有 `[N]>&-` 都会隐藏可见输出并进入 `output-suppressed`，与当前已区分 `0>&-` / `3>&-` 的实现不一致，方便的话请一并更新。

File: src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs [hunk 7]
```diff
@@ -122,6 +127,87 @@ pub(super) fn parse_command(command: &str) -> ParsedCommand {
...
+                        // redirection-word position: both bash and zsh
+                        // parse `: >&1{` as a redirection to a file named
+                        // `1{` (PR #2069 review P1), so they must reject
+                        // the elision and fail closed.
+                        matches!(
+                            next,
+                            ' ' | '\t' | '\n' | ';' | '|' | '&' | '<' | '>' | '(' | ')'
+                        )
+                    });
+                    if (has_digit || has_dash) && boundary_ok {
```
Comment (kongche-jbw) [inline src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs:196]: [P1] 想请再确认一下任意数字目标 fd 的状态假设。cosh-shell 会把批准命令发送到持久的前台 shell PTY，因此辅助描述符可以由之前的命令保留。例如先执行 `exec 3>out`，再执行 `printf payload >&3`，bash 和 zsh 都会把 `payload` 写入文件 `out`；但当前 probe 会把 `>&3` 当作普通 fd duplication 消除，使它从基线的 High 降为非 High。`2>&10` 也有同样问题：只有 fd 10 未打开时才不会落盘；若它此前已绑定到文件，后续输出会写入该文件。

考虑到解析器无法知道当前 fd 的绑定状态，是否可以只豁免目标 fd 1/2，将非标准目标 fd 继续 fail-closed（或仅在能证明绑定安全时放行），并补充 `>&3` / `2>&10` 的持久状态回归？

**Review by kongche-jbw — CHANGES_REQUESTED** [review]
感谢继续修订，多位源前缀的上一项 P1 已正确闭环。继续检查持久前台 shell 的描述符状态时，发现目标 fd 仍有一个状态相关边界，可能需要再收紧；具体见行内 P1。

另外，此前提到的 [P2] 文档同步仍未完成：PR body 仍将所有 `[N]>&-` 写成会隐藏可见输出；最新提交说明还称 `2>&10` “neither shell touches the filesystem”，但该结论只在目标 fd 未绑定到文件时成立。方便的话请一并更新。

File: src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs [hunk 8]
```diff
@@ -122,6 +127,101 @@ pub(super) fn parse_command(command: &str) -> ParsedCommand {
...
+                        .peek()
+                        .is_some_and(|next| next.is_ascii_digit())
+                    {
+                        target_digit = dup_lookahead.next();
+                        dup_consumed += 1;
+                    }
+                    // Only fd 1 and fd 2 are safe duplication targets; a
+                    // second digit (`2>&10`) or any other digit (`>&3`)
+                    // rejects the elision.
+                    let has_digit = matches!(target_digit, Some('1') | Some('2'))
```
Comment (kongche-jbw) [inline src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs:169]: [P2] 想再请教一下，这里的 “fd 1/2 是 state-independent target” 是否由 handoff / shell-host 层显式保证。这个场景在普通交互中确实不算常见，而且原始 `>` 的 hard gate 仍会让命令走 AskUser，因此我倾向把它作为非阻断的边界确认，而不是自动执行问题。

不过在持久 shell 中，fd 1/2 本身也可以由先前命令重绑定。例如我在隔离目录对 bash 和 zsh 都验证了：

```sh
exec 2>out
printf payload >&2
```

两种 shell 最终都会得到内容为 `payload` 的 `out`。fd 1 也类似：`exec 1>out` 后执行 `ls /not-found 2>&1`，错误输出会写入 `out`。这意味着当前版本会把一个实际可能写文件的形态从基线 High 降为非 High，只是触发条件较低频。

如果 shell-host 在每次批准命令前会恢复或验证 fd 1/2 指向 PTY，方便的话能否指出对应不变量并补一个回归；如果没有这层保证，是否需要在说明中明确这个状态假设，或重新评估这里的豁免边界？

**Review by kongche-jbw — COMMENTED** [review]
继续复核时补充一个低频、非阻断的 fd 状态边界，主要想确认执行层是否已经保证 fd 1/2 始终绑定到 PTY；具体 PoC 和考虑见行内 P2。

File: src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs [hunk 9]
```diff
@@ -122,6 +127,113 @@ pub(super) fn parse_command(command: &str) -> ParsedCommand {
                 if !fd_candidate {
                     push_token(&mut tokens, &mut token, &mut token_quoted);
                 }
+                // Fd duplication probe (issue #2054, spec
+                // `shell-fd-dup-redirection-risk`): only cross-shell
+                // consistent, state-independent forms are elided —
```
Comment (kongche-jbw) [inline src/cosh-ng/crates/cosh-shell/src/tools/command_risk_parser.rs]: [P2] 感谢进一步核查 shell-host；“重绑定本身已由 High 审批、且一旦重绑定普通输出命令同样会写入文件”这个策略取舍我可以接受。不过相关文档措辞似乎还没有完全同步：这里和 PR body / 提交正文仍将 fd 1/2 称为 `state-independent`，紧接着又明确说明它们可以被持久重绑定；下方的 “neither touches the filesystem” 也容易被理解为无条件结论。此外文件顶部仍泛写 `[N]>&M`，而实现实际只豁免目标 fd 1/2。

是否可以考虑统一改成类似“按策略豁免 conventional output targets 1/2；其重绑定需要先经过用户批准，且重绑定后的输出状态无法由本词法分类器防御”，并同步 PR body 与提交正文？当前这段 round-by-round 说明也比较长，若详细权衡已经记录在 design doc，源码内保留简短的不变量/策略理由可能更便于后续维护；如果该 design doc 是外部材料，方便的话也请附一下可审查的位置。

**Review by kongche-jbw — COMMENTED** [review]
感谢说明，fd 1/2 的行为边界和审批前提我认为已经解释清楚，这一项不再作为行为问题阻断。还剩一处非阻断的 [P2] 文档一致性需要同步，具体见行内评论。

**Review by kongche-jbw — APPROVED** [review]
lgtm thx
